### ISSUE TITLE
Incorrect type inference with `__radd__` with subclass of `tuple[int, ...]`

### ISSUE THREAD

**Issue by randolf-scholz** [python/mypy#19006]
**Bug Report**

`mypy` generally correctly prioritizes `__radd__` if the right operand is a subtype of the left operand. However, I discovered that it can fail to do so when creating a subclass of `tuple[int, ...]`.

**To Reproduce**

- https://mypy-play.net/?mypy=1.15.0&python=3.12&gist=720172b4776456220b93a7d4ab4af4db
- https://mypy-play.net/?mypy=1.15.0&python=3.12&gist=df9688d48ce1cea13d8bdd55b9ee1471&enable-incomplete-feature=PreciseTupleTypes

```python
from typing import assert_type

class Size(tuple[int, ...]):
    def __add__(self, other: tuple[int, ...], /) -> "Size": return Size()  # type: ignore[override]
    def __radd__(self, other: tuple[int, ...], /) -> "Size": return Size()

tup0: tuple[()] = ()
tup1: tuple[int] = (1,)
tup2: tuple[int, int] = (1, 2)
tupN: tuple[int, ...] = (1, 2, 3)
size: Size = Size([3, 4])

# __radd__
assert_type(tup0 + tupN, tuple[int, ...])  # ✅
assert_type(tup1 + tupN, tuple[int, ...])  # ✅
assert_type(tup2 + tupN, tuple[int, ...])  # ✅
assert_type(tupN + tupN, tuple[int, ...])  # ✅
assert_type(tup0 + size, Size)  # ❌ False positive
assert_type(tup1 + size, Size)  # ❌ False positive
assert_type(tup2 + size, Size)  # ❌ False positive
assert_type(tupN + size, Size)  # ✅
```

The bug does seem to be tuple-specific, for instance it does not appear with integer literals: https://mypy-play.net/?mypy=latest&python=3.12&gist=da0763e25cd0654d1a8b8b0b67291bc5

**Expected Behavior**

All `assert_type` in the example above should succeed.

**Comment by sterliakov** [comment]
Ohh, this is interesting! `tuple.__{r,}add__` methods are special-cased to handle their sizes and non-homogeneous item types. There must be something missing in fallback handling here:

https://github.com/python/mypy/blob/c724a6a806655f94d0c705a7121e3d671eced96d/mypy/checkexpr.py#L3487-L3516

**Comment by sterliakov** [comment]
That's correct: only `tuple[Any, ...]` is a subtype of all tuples including fixed-sizes one. `tuple[int, ...]` is **not** a subtype of `tuple[int, int]`, for example (while the reverse is true - a fixed-size `tuple[int, int]` is a subtype of variadic `tuple[int, ...]`). This Any special-case is intentional (because variadic Any tuples can arise from some untyped constructs), but it does not generalize to other tuples. https://mypy-play.net/?mypy=master&python=3.12&flags=strict&gist=d60b71630f447223ac70c69f8b7d7d62

Sorry, I linked to the wrong place - that one is for cases without `__radd__`, so with custom `__radd__` we fall back to plain `self.check_op`. Runtime dispatch doesn't care about tuple params, so we ignore `__radd__` when we shouldn't. Maybe just allow variadic tuples explicitly when gathering candidates in `check_op_reversible` (or one more flag for `is_subtype`)?

**Comment by randolf-scholz** [comment]
Right, so it doesn't really have anything to do with tuple literals after all, I will update the OP.

**Comment by randolf-scholz** [comment]
@sterliakov To recapitulate:

https://github.com/python/mypy/blob/d68ea3549c51c4b224f8dbdf44558df8af523e91/mypy/checkexpr.py#L4050-L4062

In the specific case here, the LHS is a fixed size tuple `tuple[*Ts]`, whereas the RHS is a variable size tuple `tuple[T, ...]`.

So `mypy` checks whether `is_subtype(tuple[int, ...], tuple[()])`, which fails.

But isn't the `is_subtype` check too strong? I may be wrong, but doesn't python at runtime simply check whether `isinstance(right, type(left))` succeeds in order to determine whether to call `__radd__`?

And a `tuple[*Ts]` is an instance of `tuple[T, ...]` if and only if `Union[*Ts] <: T`.

In our examples, this union is either empty (`Union[()] = Never`) or `int`, so everything checks out.

**Comment by randolf-scholz** [comment]
@sterliakov To recapitulate:

https://github.com/python/mypy/blob/d68ea3549c51c4b224f8dbdf44558df8af523e91/mypy/checkexpr.py#L4050-L4062

In the specific case here, the LHS is a fixed size tuple `tuple[*Ts]`, whereas the RHS is a variable size tuple `tuple[T, ...]`.

So `mypy` checks whether `is_subtype(tuple[T, ...], tuple[*Ts])`; in our specific case `is_subtype(tuple[int, ...], tuple[()])`, which obviously fails.

However, wouldn't it make sense here to consider reinterpreting the fixed size tuple as a variable size tuple?

One could ask: Does there exist a choice of `S` for which `tuple[*Ts] <: tuple[S, ...]` and `tuple[T, ...] <: tuple[S, ...]`? Indeed, this is trivially true for `S = object`, and the minimal type that satisfies these constraints is `S = T | Union[*Ts]`.

**Comment by randolf-scholz** [comment]
> Not sure I follow, are you trying to convince me that variadic tuples should be considered subtypes of some fixed-length tuples?

That was definitely not the goal 😅. The issue is that the `is_subtype` check is likely too strong to be compatible with runtime, so I was wondering if one could weaken this check - in the specific case of checking reverse-op applicability - in one way or another without breaking things.

Looking at the python C-source, the actual check performed is `PyType_IsSubtype(Py_TYPE(left), Py_TYPE(right)))` [^1], which simply checks whether `type(left) in type(right).__mro__` [^2].

This check ignores generics, hence the suggestion above was to slightly weaken the `is_subtype` check[^3] to consider upcasting the parametrization of a generic type (e.g. from `tuple[*Ts]` to `tuple[T | Union[*Ts], ...]` in this case.) I do not think it ever makes sense to upcast the base type itself (e.g. from `tuple` to `object`.)


[^1]: https://github.com/python/cpython/blob/08d7687094c6acb8c2ea1925a292a94ce1246c82/Objects/abstract.c#L919-L980
[^2]: https://github.com/python/cpython/blob/ff4959b6b0cf6da5d2f59260a9103b31c812b0ba/Objects/typeobject.c#L2695-L2724
[^3]: Of course only in the context of checking whether to apply the reverse op.

**Comment by sterliakov** [comment]
Ough, sorry, I was missing the third footnote in your previous comment - I thought you're trying to change `is_subtype` behaviour for all purposes.

Yes, at runtime we have essentially an `issubclass` check, and it really doesn't play well with static generics. Perhaps we should replace `is_subtype` with `covers_at_runtime` to determine whether `__rOP__` should be called first?

**Comment by randolf-scholz** [comment]
@sterliakov I created a draft PR, but there are still some things unclear #19046
